File: backend/src/services/prediction_service.py
"""
Prediction Service for Dasha periods.
Integrates with the Dasha engine to provide detailed predictions.
"""

from datetime import datetime
from typing import Dict, List, Optional, Any

from ..core.ephemeris import get_planet_positions
from ..core.dasha import VimshottariDasha
from ..core.predictions import DashaPredictionEngine, DashaPrediction
from ..models.schemas import BirthData


class PredictionService:
    """Service for generating Dasha predictions."""
    
    def __init__(self):
        self.prediction_engine = DashaPredictionEngine()
    
    def _get_moon_longitude(self, birth_data: BirthData) -> float:
        """Get Moon's longitude from birth data."""
        positions = get_planet_positions(
            dt=birth_data.date,
            latitude=birth_data.latitude,
            longitude=birth_data.longitude,
            timezone=birth_data.timezone,
            ayanamsa=birth_data.ayanamsa
        )
        return positions['MOON'].longitude
    
    def get_mahadasha_prediction(self, dasha_lord: str) -> Dict[str, Any]:
        """Get detailed prediction for a Mahadasha period."""
        prediction = self.prediction_engine.generate_complete_prediction(dasha_lord)
        return self._format_prediction(prediction)
    
    def get_antardasha_prediction(self, mahadasha_lord: str, 
                                  antardasha_lord: str) -> Dict[str, Any]:
        """Get detailed prediction for an Antardasha period."""
        prediction = self.prediction_engine.generate_complete_prediction(
            mahadasha_lord, 
            antardasha_lord
        )
        return self._format_prediction(prediction)
    
    def get_pratyantardasha_prediction(self, mahadasha_lord: str,
                                       antardasha_lord: str,
                                       pratyantardasha_lord: str) -> Dict[str, Any]:
        """Get detailed prediction for a Pratyantardasha period."""
        prediction = self.prediction_engine.generate_complete_prediction(
            mahadasha_lord,
            antardasha_lord,
            pratyantardasha_lord
        )
        return self._format_prediction(prediction)
    
    def get_current_period_prediction(self, birth_data: BirthData,
                                      target_date: Optional[datetime] = None) -> Dict[str, Any]:
        """Get prediction for the current running Dasha period."""
        moon_lon = self._get_moon_longitude(birth_data)
        dasha = VimshottariDasha(moon_lon, birth_data.date)
        
        if target_date is None:
            target_date = datetime.now()
        
        current = dasha.get_current_periods(target_date)
        
        if not current:
            return {"error": "Could not determine current dasha periods"}
        
        mahadasha_lord = current['mahadasha']['lord']
        antardasha_lord = current.get('antardasha', {}).get('lord')
        pratyantardasha_lord = current.get('pratyantardasha', {}).get('lord')
        
        prediction = self.prediction_engine.generate_complete_prediction(
            mahadasha_lord,
            antardasha_lord,
            pratyantardasha_lord
        )
        
        result = self._format_prediction(prediction)
        
        # Add period information
        result['current_periods'] = {
            'mahadasha': {
                'lord': mahadasha_lord,
                'start': current['mahadasha']['start'].isoformat(),
                'end': current['mahadasha']['end'].isoformat()
            }
        }
        
        if antardasha_lord:
            result['current_periods']['antardasha'] = {
                'lord': antardasha_lord,
                'start': current['antardasha']['start'].isoformat(),
                'end': current['antardasha']['end'].isoformat()
            }
        
        if pratyantardasha_lord:
            result['current_periods']['pratyantardasha'] = {
                'lord': pratyantardasha_lord,
                'start': current['pratyantardasha']['start'].isoformat(),
                'end': current['pratyantardasha']['end'].isoformat()
            }
        
        return result
    
    def get_timeline_with_predictions(self, birth_data: BirthData,
                                      years_ahead: int = 80) -> List[Dict[str, Any]]:
        """Get complete Mahadasha timeline with predictions for each period."""
        moon_lon = self._get_moon_longitude(birth_data)
        dasha = VimshottariDasha(moon_lon, birth_data.date)
        
        timeline = dasha.generate_mahadasha_timeline()
        
        result = []
        for md in timeline:
            md_lord = md['lord']
            prediction = self.prediction_engine.generate_complete_prediction(md_lord)
            
            # Get Antardasha periods
            antardashas = dasha.calculate_antardasha(md)
            antardasha_list = []
            
            for ad in antardashas:
                ad_lord = ad['lord']
                ad_prediction = self.prediction_engine.generate_complete_prediction(
                    md_lord, ad_lord
                )
                
                antardasha_list.append({
                    'lord': ad_lord,
                    'start': ad['start'].isoformat(),
                    'end': ad['end'].isoformat(),
                    'days': round(ad['days'], 1),
                    'prediction_summary': {
                        'overall_theme': ad_prediction.overall_theme,
                        'overall_rating': ad_prediction.overall_rating,
                        'health': {
                            'trend': ad_prediction.predictions['health'].trend,
                            'summary': ad_prediction.predictions['health'].summary
                        },
                        'wealth': {
                            'trend': ad_prediction.predictions['wealth'].trend,
                            'summary': ad_prediction.predictions['wealth'].summary
                        },
                        'career': {
                            'trend': ad_prediction.predictions['career'].trend,
                            'summary': ad_prediction.predictions['career'].summary
                        },
                        'relationships': {
                            'trend': ad_prediction.predictions['relationships'].trend,
                            'summary': ad_prediction.predictions['relationships'].summary
                        },
                        'favorable_activities': ad_prediction.favorable_activities[:3],
                        'remedies': {
                            'gemstone': ad_prediction.gemstone,
                            'mantra': ad_prediction.mantra,
                            'deity': ad_prediction.deity
                        }
                    }
                })
            
            result.append({
                'lord': md_lord,
                'start': md['start'].isoformat(),
                'end': md['end'].isoformat(),
                'years': round(md['years'], 2),
                'is_birth_dasha': md.get('is_birth_dasha', False),
                'prediction': self._format_prediction(prediction),
                'antardashas': antardasha_list
            })
        
        return result
    
    def _format_prediction(self, prediction: DashaPrediction) -> Dict[str, Any]:
        """Format a DashaPrediction into a dictionary."""
        formatted_predictions = {}
        
        for area, pred in prediction.predictions.items():
            formatted_predictions[area] = {
                'trend': pred.trend,
                'intensity': pred.intensity,
                'summary': pred.summary,
                'details': pred.details,
                'remedies': pred.remedies,
                'keywords': pred.keywords
            }
        
        return {
            'dasha_lord': prediction.dasha_lord,
            'period_type': prediction.period_type,
            'overall_theme': prediction.overall_theme,
            'overall_rating': prediction.overall_rating,
            'predictions': formatted_predictions,
            'favorable_activities': prediction.favorable_activities,
            'unfavorable_activities': prediction.unfavorable_activities,
            'important_transits': prediction.important_transits,
            'remedies': {
                'gemstone': prediction.gemstone,
                'mantra': prediction.mantra,
                'deity': prediction.deity
            }
        }
    
    def get_all_dasha_predictions(self) -> Dict[str, Dict[str, Any]]:
        """Get predictions for all 9 Dasha lords."""
        dasha_lords = ['Sun', 'Moon', 'Mars', 'Mercury', 'Jupiter', 
                       'Venus', 'Saturn', 'Rahu', 'Ketu']
        
        return {
            lord: self.get_mahadasha_prediction(lord)
            for lord in dasha_lords
        }
    
    def get_combination_matrix(self) -> Dict[str, Dict[str, Dict[str, Any]]]:
        """Get prediction matrix for all Mahadasha-Antardasha combinations."""
        dasha_lords = ['Sun', 'Moon', 'Mars', 'Mercury', 'Jupiter',
                       'Venus', 'Saturn', 'Rahu', 'Ketu']
        
        matrix = {}
        for md_lord in dasha_lords:
            matrix[md_lord] = {}
            for ad_lord in dasha_lords:
                prediction = self.prediction_engine.generate_complete_prediction(
                    md_lord, ad_lord
                )
                # Include just summary for the matrix
                matrix[md_lord][ad_lord] = {
                    'overall_theme': prediction.overall_theme,
                    'overall_rating': prediction.overall_rating,
                    'health_trend': prediction.predictions['health'].trend,
                    'wealth_trend': prediction.predictions['wealth'].trend,
                    'career_trend': prediction.predictions['career'].trend,
                    'relationships_trend': prediction.predictions['relationships'].trend,
                }
        
        return matrix

File: frontend/src/lib/services/predictionService.ts
/**
 * Prediction Service — local-only, delegates to DashaPredictionEngine
 */

import { getPlanetPositions, type PlanetPosition } from '../core/ephemeris';
import { VimshottariDasha } from '../core/dasha';
import { DashaPredictionEngine, type DashaPrediction, type ChartContext } from '../core/predictions';
import { type Lang, getStoredLang } from '../core/i18n';
import { computeAshtakavarga, type Contributor } from '../core/ashtakavarga';
import { getCurrentTransits, summarizeGocharaForPrediction } from '../core/transits';
import type { BirthData } from '../../types/astrology';
import type { DashaPredictionData } from '../../services/api';

function toIso(d: Date): string { return d.toISOString(); }

export function formatPrediction(pred: DashaPrediction): DashaPredictionData {
  const predictions = Object.fromEntries(
    Object.entries(pred.predictions).map(([area, p]) => [area, {
      trend: p.trend, intensity: p.intensity, summary: p.summary,
      details: p.details, remedies: p.remedies, keywords: p.keywords,
    }])
  ) as DashaPredictionData['predictions'];

  return {
    dashaLord: pred.dashaLord,
    antardasha: pred.antardasha,
    pratyantardasha: pred.pratyantardasha,
    sookshmaDasha: pred.sookshmaDasha,
    periodType: pred.periodType,
    overallTheme: pred.overallTheme,
    overallRating: pred.overallRating,
    predictions,
    favorableActivities: pred.favorableActivities,
    unfavorableActivities: pred.unfavorableActivities,
    importantTransits: pred.importantTransits,
    lordStrengths: pred.lordStrengths,
    natalFoundation: pred.natalFoundation,
    remedies: { gemstone: pred.gemstone, mantra: pred.mantra, deity: pred.deity },
    combinationWarning: pred.combinationWarning,
    combinationBonus: pred.combinationBonus,
  };
}

async function getMoonLon(bd: BirthData): Promise<number> {
  const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
  return positions['MOON'].longitude;
}

const ENGINE_PLANET_KEY: Record<string, Contributor> = {
  SUN: 'Sun', MOON: 'Moon', MARS: 'Mars', MERCURY: 'Mercury',
  JUPITER: 'Jupiter', VENUS: 'Venus', SATURN: 'Saturn',
};

export function buildChartContext(positions: Record<string, PlanetPosition>): ChartContext {
  const rashis = { Lagna: positions['ASCENDANT'].rashi } as Record<Contributor, number>;
  for (const [k, v] of Object.entries(ENGINE_PLANET_KEY)) {
    rashis[v] = positions[k].rashi;
  }
  const ascRashi = positions['ASCENDANT'].rashi;
  // Whole-sign natal house + full natal condition for every body
  // (including the nodes, which Ashtakavarga skips).
  const planetHouses: Record<string, number> = {};
  const planetRashis: Record<string, number> = {};
  const planetLongitudes: Record<string, number> = {};
  const planetRetro: Record<string, boolean> = {};
  for (const key of ['SUN','MOON','MARS','MERCURY','JUPITER','VENUS','SATURN','RAHU','KETU']) {
    const p = positions[key];
    if (p == null) continue;
    const name = key.charAt(0) + key.slice(1).toLowerCase(); // SUN -> Sun
    planetHouses[name] = ((p.rashi - ascRashi + 12) % 12) + 1;
    planetRashis[name] = p.rashi;
    planetLongitudes[name] = p.longitude;
    planetRetro[name] = p.isRetrograde;
  }
  return {
    ashtakavarga: computeAshtakavarga(rashis),
    planetHouses,
    ascendantRashi: ascRashi,
    planetRashis,
    planetLongitudes,
    planetRetro,
    moonRashi: positions['MOON'].rashi,
  };
}

async function getCurrentDashaContext(bd: BirthData): Promise<{
  moonLon: number; ctx: ChartContext;
}> {
  const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
  return { moonLon: positions['MOON'].longitude, ctx: buildChartContext(positions) };
}

const engine = new DashaPredictionEngine();

async function chartCtxFor(bd: BirthData): Promise<ChartContext> {
  const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
  return buildChartContext(positions);
}

export async function getMahadashaPrediction(bd: BirthData, dashaLord: string, lang: Lang = getStoredLang()): Promise<DashaPredictionData> {
  const ctx = await chartCtxFor(bd);
  return formatPrediction(engine.generateCompletePrediction(dashaLord, undefined, undefined, undefined, ctx, lang));
}

export async function getAntardashaPrediction(bd: BirthData, mahadasha: string, antardasha: string, lang: Lang = getStoredLang()): Promise<DashaPredictionData> {
  const ctx = await chartCtxFor(bd);
  return formatPrediction(engine.generateCompletePrediction(mahadasha, antardasha, undefined, undefined, ctx, lang));
}

export async function getPratyantardashaPrediction(bd: BirthData, mahadasha: string, antardasha: string, pratyantardasha: string, lang: Lang = getStoredLang()): Promise<DashaPredictionData> {
  const ctx = await chartCtxFor(bd);
  return formatPrediction(engine.generateCompletePrediction(mahadasha, antardasha, pratyantardasha, undefined, ctx, lang));
}

/** Just the Ashtakavarga grid for the chart. */
export async function getAshtakavargaForChart(bd: BirthData) {
  const positions = await getPlanetPositions(bd.date, bd.latitude, bd.longitude, bd.timezone, bd.ayanamsa);
  return computeAshtakavarga({
    Lagna: positions['ASCENDANT'].rashi,
    Sun: positions['SUN'].rashi,
    Moon: positions['MOON'].rashi,
    Mars: positions['MARS'].rashi,
    Mercury: positions['MERCURY'].rashi,
    Jupiter: positions['JUPITER'].rashi,
    Venus: positions['VENUS'].rashi,
    Saturn: positions['SATURN'].rashi,
  });
}

export async function getCurrentPeriodPrediction(bd: BirthData, targetDate?: Date, lang: Lang = getStoredLang()): Promise<DashaPredictionData> {
  const { moonLon, ctx } = await getCurrentDashaContext(bd);
  const calc = new VimshottariDasha(moonLon, new Date(bd.date));
  const td = targetDate ?? new Date();
  const current = calc.getCurrentPeriods(td);
  if ('error' in current) throw new Error(current.error);

  // Fold current transits (Sade Sati, Guru blessing…) into the prediction.
  try {
    const gochara = await getCurrentTransits(bd.ayanamsa, ctx.moonRashi!, ctx.ascendantRashi!, td, undefined, undefined, lang);
    const transitSummary = summarizeGocharaForPrediction(gochara, lang);
    ctx.transitNotes = transitSummary.notes;
    ctx.transitScoreMod = transitSummary.scoreMod;
  } catch {
    // Transits are an enhancement — never block the core prediction.
  }

  const pred = engine.generateCompletePrediction(
    current.mahadasha.lord,
    current.antardasha.lord,
    current.pratyantardasha?.lord,
    current.sookshmaDasha?.lord,
    ctx,
    lang,
  );
  const result = formatPrediction(pred);
  result.currentPeriods = {
    mahadasha: { lord: current.mahadasha.lord, start: toIso(current.mahadasha.start), end: toIso(current.mahadasha.end) },
    antardasha: { lord: current.antardasha.lord, start: toIso(current.antardasha.start), end: toIso(current.antardasha.end) },
    ...(current.pratyantardasha ? {
      pratyantardasha: { lord: current.pratyantardasha.lord, start: toIso(current.pratyantardasha.start), end: toIso(current.pratyantardasha.end) },
    } : {}),
    ...(current.sookshmaDasha ? {
      sookshmaDasha: { lord: current.sookshmaDasha.lord, start: toIso(current.sookshmaDasha.start), end: toIso(current.sookshmaDasha.end) },
    } : {}),
  };
  return result;
}

export interface SookshmaEntry {
  lord: string;
  start: string;
  end: string;
  days: number;
  isCurrent: boolean;
}

export interface SookshmaPeriodList {
  mahadasha: string;
  antardasha: string;
  pratyantardasha: string;
  pratyantarStart: string;
  pratyantarEnd: string;
  sookshmas: SookshmaEntry[];
  currentSookshmaLord: string | null;
}

export async function getSookshmaPeriodsForCurrent(bd: BirthData, targetDate?: Date): Promise<SookshmaPeriodList | null> {
  const moonLon = await getMoonLon(bd);
  const calc = new VimshottariDasha(moonLon, new Date(bd.date));
  const td = targetDate ?? new Date();
  const current = calc.getCurrentPeriods(td);
  if ('error' in current || !current.pratyantardasha) return null;

  const mahadashas = calc.generateMahadashaTimeline();
  const currentMd = mahadashas.find(md => md.lord === current.mahadasha.lord && md.start <= td && md.end >= td);
  if (!currentMd) return null;
  const antardashas = calc.calculateAntardasha(currentMd);
  const currentAd = antardashas.find(ad => ad.lord === current.antardasha.lord && ad.start <= td && ad.end >= td);
  if (!currentAd) return null;
  const pratyantars = calc.calculatePratyantardasha(currentAd);
  const currentPd = pratyantars.find(pd => pd.lord === current.pratyantardasha!.lord && pd.start <= td && pd.end >= td);
  if (!currentPd) return null;

  const sookshmas = calc.calculateSookshmaDasha(currentPd);

  return {
    mahadasha: current.mahadasha.lord,
    antardasha: current.antardasha.lord,
    pratyantardasha: current.pratyantardasha.lord,
    pratyantarStart: toIso(currentPd.start),
    pratyantarEnd: toIso(currentPd.end),
    currentSookshmaLord: current.sookshmaDasha?.lord ?? null,
    sookshmas: sookshmas.map(sd => ({
      lord: sd.lord,
      start: toIso(sd.start),
      end: toIso(sd.end),
      days: Math.round(sd.days * 10) / 10,
      isCurrent: sd.lord === current.sookshmaDasha?.lord &&
        new Date(toIso(sd.start)) <= td && new Date(toIso(sd.end)) >= td,
    })),
  };
}

export async function getTimelineWithPredictions(bd: BirthData, yearsAhead = 80, lang: Lang = getStoredLang()): Promise<unknown[]> {
  const { moonLon, ctx } = await getCurrentDashaContext(bd);
  const calc = new VimshottariDasha(moonLon, new Date(bd.date));
  const mahadashas = calc.generateMahadashaTimeline(yearsAhead);

  return mahadashas.map(md => {
    const prediction = engine.generateCompletePrediction(md.lord, undefined, undefined, undefined, ctx, lang);
    const antardashas = calc.calculateAntardasha(md);

    return {
      lord: md.lord,
      start: toIso(md.start),
      end: toIso(md.end),
      years: Math.round(md.years * 100) / 100,
      isBirthDasha: md.isBirthDasha,
      prediction: formatPrediction(prediction),
      antardashas: antardashas.map(ad => {
        const adPred = engine.generateCompletePrediction(md.lord, ad.lord, undefined, undefined, ctx, lang);
        return {
          lord: ad.lord,
          start: toIso(ad.start),
          end: toIso(ad.end),
          days: Math.round(ad.days * 10) / 10,
          predictionSummary: {
            overallTheme: adPred.overallTheme,
            overallRating: adPred.overallRating,
            health: { trend: adPred.predictions.health.trend, summary: adPred.predictions.health.summary },
            wealth: { trend: adPred.predictions.wealth.trend, summary: adPred.predictions.wealth.summary },
            career: { trend: adPred.predictions.career.trend, summary: adPred.predictions.career.summary },
            relationships: { trend: adPred.predictions.relationships.trend, summary: adPred.predictions.relationships.summary },
            favorableActivities: adPred.favorableActivities.slice(0, 3),
            remedies: { gemstone: adPred.gemstone, mantra: adPred.mantra, deity: adPred.deity },
            combinationWarning: adPred.combinationWarning,
            combinationBonus: adPred.combinationBonus,
          },
        };
      }),
    };
  });
}
